### ISSUE TITLE
Embeds Taking Long Time To Load Now (Discord)

### ISSUE THREAD

**Issue by snnacks** [Hubro/clipface#20]
I am not sure if this is related to Clipface, but recently something has gone a little weird with the embedding. Now, the longer a clip is, the longer it takes for Discord to load the embed player. Sometimes, if the clip is too long, Discord won't even load it and will show the poop error image. I was using Cloudflare to get a free SSL certificate and I even tried bypassing their cache, but that isn't solving the issue. I was wondering if this is related to me having more and more clips in the clipface directory slowing it down or is this just a Discord thing? I almost feel like Discord's servers are trying to load the entire video before loading the embed player.

Example: https://imgur.com/COeY16O

EDIT: This honestly just might be random. A 34 second clip also generates the poop error image on Discord and won't bring up an embed player.

**Comment by Hubro** [comment]
I'm not sure if this could be Clipface's fault, all we do is just expose a few meta tags in the HTML document that point to the video file. Would you be able to do some experiments? Upload the relevant clips to Streamable and try to embed them, see if that gives you the same result?

**Comment by Snnackss** [comment]
Streamable has no issues. I blame my Cloudflare proxy. I am switching VPS providers because my needs for a VPS have gone from just Clipface to several other services/websites. So, when I set it up I will use let's encrypt to generate SSL certificates and won't use Cloudflare's proxy and then I'll see if that solves my issue. Since I am doing that very soon I am going to leave this issue open until I see if it's Cloudflare (In case other people use their proxy for clipface in the future).

**Comment by Brando47** [comment]
Been following this issue as I'm getting the same (only just setup clipface last night). Not running mine behind cloudflare or anything, I do only have about 16mbps up though. Works fine for smaller files, I agree it feels like it tries to load the whole video or something. Happy to help with diagnosing if there's any ideas on how to chase it up.

Big thanks for the software btw, does one thing and does it really well!

**Comment by longdas** [comment]
I am having a similiar issue on gigabit upload. Knowing Discord, it's probably their fault 😆

**Comment by Snnackss** [comment]
I just setup Clipface on my new VM. Discord is doing the same thing. I feel like they are scanning the file for viruses and since my domain is not "trusted" it isn't instantly loaded like Streamable or something. This is a huge bummer now because some embeds don't even load...

**Comment by Snnackss** [comment]
I looked into this some more. Based on my testing it seems like anything over 100MB refuses to load as an embed and I feel like this has to do with Discord's 100MB upload limit. Lately Discord has been updating Nitro like crazy and the expensive Nitro unlocks the 100MB upload limit. I would not be surprised if they did some back end changes and now embeds have been thrown into this requirement because Discord's servers proxies all embedded video clips.

So, basically, keep it down to under 100MB and any clips you expect to be larger than that, upload them to YouTube or something if you want them to be embedded.

NOTE: I don't have any true sources to confirm this, but I am very convinced this is what is going on.

**Comment by Snnackss** [comment]
I am going to close this issue now. Hubro, you can look into this more if you want, but you may want to put that into the README.md about how anything beyond 100MB on Discord won't embed anymore.

**Comment by Hubro** [comment]
@Snnackss https://github.com/Hubro/clipface/commit/c2550d41949b107f69e134a9a89aa3967720ad8b

**Comment by chunguscodes** [comment]
This isn't the issue @Hubro - Discord does not limit the size of embedded videos as they are not proxied, it loads directly from the source. I've tested this personally with a 2GB file and it plays as expected, it's the same embed code that Streamable.com also uses. The below would mean you'd have to do some processing like generating a thumbnail etc but if you don't want to do that, you can just hardcode a single thumbnail or something. The loltf variables are personal ones I used for my software, and can be changed.

The following opengraph metadata should work:

```
<meta property="og:type" content="video">
<meta property="og:url" content="${loltfurl}">
<meta property="og:video" content="${loltfurl}">
<meta property="og:image" content="https://generatedthumbnail.com">
<meta property="og:video:url" content="${loltfurl}">
<meta property="og:video:secure_url" content="${loltfurl}">
<meta property="og:video:type" content="${f.mime}">
<meta property="og:video:width" content="600">
<meta property="og:video:height" content="400">
<meta property="og:title" content="titleofvideo">

**Comment by Hubro** [comment]
@max4186 The only difference there is that clipface isn't specifying `og:image` :thinking: Do you mean that lacking an image is the reason large videos don't load in Discord? I'm not sure how that makes sense, why would it only happen on large videos? Have you tested this hypothesis?

**Comment by ghost** [comment]
Hey Hubro, a buddy of mine and I forked Clipface to mess around with it a bit to have some fun, but one of the things we did was test max's theory about the `og:image`. Turns out max is correct. I tested a 150MB clip and it didn't load, then we just forced every video to use the same test image and restarted clipface with that change. Then, the embed instantly loaded, but with the placeholder image there and now the 150MB file plays in Discord embedded.

[Here is the commit on our fork where we made that change](https://github.com/Hubro/clipface/commit/9d230e1bbfd82afe4334a1c21d46db0a2bdb6ed7)

**Comment by Hubro** [comment]
> Hey Hubro, a buddy of mine and I forked Clipface to mess around with it a bit to have some fun, but one of the things we did was test max's theory about the `og:image`. Turns out max is correct. I tested a 150MB clip and it didn't load, then we just forced every video to use the same test image and restarted clipface with that change. Then, the embed instantly loaded, but with the placeholder image there and now the 150MB file plays in Discord embedded.
> 
> [Here is the commit on our fork where we made that change](https://github.com/Hubro/clipface/commit/9d230e1bbfd82afe4334a1c21d46db0a2bdb6ed7)

Interesting! So I guess the reason Discord downloads the video is in order to generate a thumbnail image.

I guess I have to add thumbnail generation 😄
